### DISCUSSION TITLE
Is there api to fetch input/output description in c?

### DISCUSSION THREAD

**Discussion by lifang-zhang** [microsoft/onnxruntime#13752] [Other Q&A]
Like the marked information in following picture.

![image](https://user-images.githubusercontent.com/86099701/203999688-713d1390-09a2-43db-b41a-57ac50ba2fab.png)

I want the [N, C, H, W] information there.

**Comment by mortenwp** [comment]
I had a similar question. Basically how do I get the ordering of the data.

I have tried to use to GetDenotationFromTypeInfo in the C api (see code below), but currently I just get the empty string
from that call. 

From what I understand of the denotation the ordering of channel, data dimension, and batch dimension are/can be tracked in onnxruntime through  the whole graph, so if the user by accident flips the dimension in a wrong way it can be detected... but that is a bit incompatible with that my char* ptr is not being populated

My models has been converted from tensorflow using tf2onnx. But not sure if the denotation info actually are in my onnx file

I don't know if this is of any help  :)

best regards
Morten

```
std::string get_input_denotation(std::size_t idx) const
{
           Ort::TypeInfo typ_info = m_session.GetInputTypeInfo(idx);
           const char* ptr;
           std::size_t len;
           const OrtApi* ort = OrtGetApiBase()->GetApi(ORT_API_VERSION);
           OrtStatus* stat(ort->GetDenotationFromTypeInfo(typ_info,&ptr,&len));
           if(stat==nullptr)
            return std::string(ptr);
          return std::string();
}
```

**Reply by lifang-zhang** [reply]
Thanks! Didn't notice function `GetDenotationFromTypeInfo`. This is useful since most of the models indicate the NCHW info in dimention denotation, as proposed [here](https://github.com/onnx/onnx/blob/main/docs/TypeDenotation.md). 

In the model ([fns-candy.onnx](https://github.com/microsoft/Windows-Machine-Learning/blob/master/Samples/CustomTensorization/CustomTensorization/fns-candy.onnx)) I used, the string is in the `doc_string` of input node, so can't use `GetDenotationFromTypeInfo`.

BTW, if you are not sure if the denotation info actually are in my onnx file, you can use python to check it out. For example, I used following code to check the input `doc_string`. So in your case, you can use `print(onnx_model.graph.input[0])` or `print(onnx_model.graph.input[0].type.tensor_type)` to see what the dimension denotation is, if it is empty there, then it is not the problem of onnxruntime.

```
from onnx import load

with open("candy.onnx", "rb") as f:
    onnx_model = load(f)

print(onnx_model.graph.input[0].doc_string)

```

**Reply by lifang-zhang** [reply]
Oh, I might be wrong, I am guessing `GetDenotationFromTypeInfo` is to get denotation of type instead of dimention, so can use `print(onnx_model.graph.input[0].type)` to check if there is any denotation.

**Reply by mortenwp** [reply]
Thanks, I will try that. I have asked on the tf2onnx project if denotation is supported when they convert from tf or keras to onnx.
I will try the python route.

There seams to be a whole set of ModelMetaDataGet functions  i.e. ModelMetaDataGetGraphDescription in the C-api. I don't know if one of these could be used to extract the information from your model?

**Reply by lifang-zhang** [reply]
Yes, I have used `ModelMetaDataGet ` to get the metadata about the model. But the `doc_string` is under graph, so it does not belong to metadata :(

And not all metadata can be fetched in onnxruntime.

![image](https://user-images.githubusercontent.com/86099701/204767520-708009d4-d044-4f6f-bf7c-782b5de109b6.png)

**Reply by lifang-zhang** [reply]
In my model `canny.onnx`, the put author and license in the metadata_props, so I fetch it through function `ModelMetadataLookupCustomMetadataMap()`

**Reply by mortenwp** [reply]
ModelMetadataGetGraphDescription says it returns the description of the graph present in the model (doc_string field of the GraphProto message within the ModelProto message). If it doesn't exist, an empty string will be returned. Isn't that what you after?
